src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/EvolutionaryAlgorithmQueryOptimizer.java

```diff
@@ -0,0 +1,143 @@
+
+        // generate candidates for the first generation
+        final RandomizedSelection<RuleApplication> ruleRandomizedSelect = new RandomizedSelection<>();
+        while ( currentGen.size() < nmCandidates && ruleApps.size() > 0 ){
+            // Pick one rule application based on priority, and apply the selected rule application to the initial plan
+            final RuleApplication ruleApplication = ruleRandomizedSelect.pickOne( ruleApps );
+            currentGen.add( ruleApplication.getResultingPlan() );
+
+            // remove the applied rule application from the set of rule applications
+            ruleApps.remove( ruleApplication );
+        }
+        ruleApplicationCache.replaceRuleApplications( plan, ruleApps);
+
+        return annotatePlansWithCost(currentGen);
+    }
+
+    protected List<PhysicalPlanWithCost> generateTheNextGen( final List<PhysicalPlanWithCost> currentGen ) throws CostEstimationException {
+        List<PhysicalPlan> newCandidates = new ArrayList<>();
+
+        while ( newCandidates.size() < nmCandidates ) {
+            // pick a parent from the current generation
+            final RandomizedSelection<PhysicalPlanWithCost> planRandomizedSelect = new RandomizedSelection<>();
+            final PhysicalPlan parent = ( planRandomizedSelect.pickOne( new HashSet<>(currentGen) )).getPlan();
+
+            // determine all possible applications of rewriting rules for the parent plan
+            Set<RuleApplication> ruleApps = ruleApplicationCache.determineRuleApplications( parent );
+
+            final RandomizedSelection<RuleApplication> ruleRandomizedSelect = new RandomizedSelection<>();
+            if ( ruleApps.size() > 0) {
+                // Pick one rewriting rule based on priority
+                final RuleApplication ruleApplication = ruleRandomizedSelect.pickOne( ruleApps );
+                newCandidates.add( ruleApplication.getResultingPlan() );
+                ruleApplicationCache.removeRuleApplications( parent, ruleApplication );
+            }
+        }
+        currentGen.addAll( annotatePlansWithCost(newCandidates) );
+
+        // select the next generation from all candidates
+        return selectNextGenFromCandidates( currentGen );
+    }
+
+    protected List<PhysicalPlanWithCost> annotatePlansWithCost( final List<PhysicalPlan> plans ) throws CostEstimationException {
+        final Double[] costs = CostEstimationUtils.getEstimates( costModel, plans );
+
+        final List<PhysicalPlanWithCost> plansWithCost = new ArrayList<>();
+        for ( int i = 0; i < plans.size(); i++ ) {
+            plansWithCost.add( new PhysicalPlanWithCost( plans.get(i), costs[i] ) );
+        }
+
+        return plansWithCost;
+    }
+
+    protected List<PhysicalPlanWithCost> selectNextGenFromCandidates( final List<PhysicalPlanWithCost> planWithCosts ) {
+        List<PhysicalPlanWithCost> currentGen = new ArrayList<>();
+
+        for ( int i = 0; i < (nmCandidates-nmSurvivors); i++ ) {
```

Comment:
You should add an `assert` check in the constructor of this class to make sure that `nmCandidates-nmSurvivors` does not result in a negative number.